assets/js/blocks/featured-items/with-featured-item.tsx

```diff
@@ -3,27 +3,103 @@
 /**
  * External dependencies
  */
+import type { BlockAlignment } from '@wordpress/blocks';
+import { ProductResponseItem } from '@woocommerce/types';
 import { __experimentalGetSpacingClassesAndStyles as getSpacingClassesAndStyles } from '@wordpress/block-editor';
 import { Icon, Placeholder, Spinner } from '@wordpress/components';
 import classnames from 'classnames';
 import { isEmpty } from 'lodash';
-import { useCallback, useState } from 'react';
+import {
+	ComponentType,
+	Dispatch,
+	SetStateAction,
+	useCallback,
+	useState,
+} from 'react';
+import { WP_REST_API_Category } from 'wp-types';
 
 /**
  * Internal dependencies
  */
 import { CallToAction } from './call-to-action';
 import { ConstrainedResizable } from './constrained-resizable';
+import { EditorBlock, GenericBlockUIConfig } from './types';
 import { useBackgroundImage } from './use-background-image';
 import {
 	dimRatioToClass,
 	getBackgroundImageStyles,
 	getClassPrefixFromName,
 } from './utils';
 
-export const withFeaturedItem = ( { emptyMessage, icon, label } ) => (
-	Component
-) => ( props ) => {
+interface WithFeaturedItemConfig extends GenericBlockUIConfig {
+	emptyMessage: string;
+}
+
+export interface FeaturedItemRequiredAttributes {
+	contentAlign: BlockAlignment;
+	dimRatio: number;
+	focalPoint: { x: number; y: number };
+	hasParallax: boolean;
+	imageFit: 'cover' | 'none';
+	isRepeated: boolean;
+	linkText: string;
+	mediaId: number;
+	mediaSrc: string;
+	minHeight: number;
+	overlayColor: string;
+	overlayGradient: string;
+	showDesc: boolean;
+	showPrice: boolean;
+}
+
+interface FeaturedCategoryRequiredAttributes
+	extends FeaturedItemRequiredAttributes {
+	categoryId: number | 'preview';
+	productId: never;
+}
+
+interface FeaturedProductRequiredAttributes
+	extends FeaturedItemRequiredAttributes {
+	categoryId: never;
+	productId: number | 'preview';
+}
+
+interface FeaturedItemRequiredProps< T > {
+	attributes: (
+		| FeaturedCategoryRequiredAttributes
+		| FeaturedProductRequiredAttributes
+	 ) &
+		EditorBlock< T >[ 'attributes' ] & {
+			// This is hardcoded because border is not yet included in Gutenberg's
```

Comment:
If we haven't yet, it might make sense to log this as an issue and/or contribute upstream.

Reply:
Yep, a few contributions upstream are on my notes.